library/std/src/sys/unix/args.rs

```diff
@@ -82,16 +77,10 @@ mod imp {
     use crate::ptr;
     use crate::sync::atomic::{AtomicIsize, AtomicPtr, Ordering};
 
-    use crate::sys_common::mutex::StaticMutex;
-
     static ARGC: AtomicIsize = AtomicIsize::new(0);
     static ARGV: AtomicPtr<*const u8> = AtomicPtr::new(ptr::null_mut());
-    // We never call `ENV_LOCK.init()`, so it is UB to attempt to
-    // acquire this mutex reentrantly!
-    static LOCK: StaticMutex = StaticMutex::new();
 
     unsafe fn really_init(argc: isize, argv: *const *const u8) {
-        let _guard = LOCK.lock();
         ARGC.store(argc, Ordering::Relaxed);
         ARGV.store(argv as *mut _, Ordering::Relaxed);
```

Comment:
This should probably be a release store to ensure the writes to the data that sits behind this `argv` pointer are visible.

Reply:
Those writes are performed before `lang_start` is called, by code that must assume that we may read from that memory immediately without atomics or synchronization.

For a spec reference, I found C11 5.1.2.2.1p2 which says the argv array is given values "prior to program startup".

Reply:
So this is directly using the system-provided argc/argv, not something constructed by Rust itself? I guess it's technically correct then. But this seems subtle enough that it warrants a comment.

I'd probably still use release/acquire accesses. They are just more obviously correct. IMO `Relaxed` should only be used when there is a good perf reason to do so, and needs to come with comments explaining why not doing release/acquire is okay. `Relaxed` is a very subtle memory order that should not be used lightly.

Reply:
Yes, this is the system-provided argc/argv. I submitted #87279 to address this.